### DISCUSSION TITLE
#Review tednaaa/paint

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/examples#14] [Sliced Review]
**Repository:** https://github.com/tednaaa/paint

**Methodology version:** 2.0

**Comment by azinit** [comment]
Так-с. Таки нашел время на проект, @tednaaa готовься))

**Comment by azinit** [comment]
## 🟩 Начну с хорошего

[`praise:`](https://conventionalcomments.org/) Очень радует, что в целом структура оч валидна, сразу плюс к +100000 соц.рейтингу FSD  😏 
![image](https://user-images.githubusercontent.com/42924400/149590231-60edf0a4-d6e9-4eed-8ae8-609569d7f1a3.png)

**Reply by azinit** [reply]
[`praise:`](https://conventionalcomments.org/) И в целом, если "окунаться вглубь", то тоже конвенциональность некая прослеживается, что значительно упрощает понимание

![image](https://user-images.githubusercontent.com/42924400/149590656-765f3b24-232d-4c61-9b03-1778e37b91d8.png)

![image](https://user-images.githubusercontent.com/42924400/149590599-e53dfaf1-7c85-43ef-af2c-8843618540a4.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Прям образцовая апишка вышла, прям как и предполагалось)

С четким разделением типов ресурсов

[`nitpick(public-api):`](https://conventionalcomments.org/) Но в то же время странно, что нет файла-реэкспорта и у нас прям в пути проявляется `shared/api/rest/image/image.ts`

![image](https://user-images.githubusercontent.com/42924400/149590782-4358009d-3b44-47f3-918f-9746d93beaad.png)

**Reply by azinit** [reply]
вот тут тоже классный пример, что инстансы для API создаются в shared/api

Особенно если потом понадобится это инициализировать поптом на уровне выше

![image](https://user-images.githubusercontent.com/42924400/149593940-80995b13-3e4e-46e2-a3a0-41a28c61b636.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Тоже выглядит достаточно унифицированно, особенно учитывая, что `color` действительно является ключевой абстракцией в приложении

![image](https://user-images.githubusercontent.com/42924400/149590976-adf7282c-b75e-4063-8dc9-1a50566f7c81.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) C одной стороны - интересно, что сразу относишь к ассетам шрифты

[`suggestion(to-improve):`](https://conventionalcomments.org/) А в целом, кажется, что можно попробовать "полуконвенциональный" формат, с папкой `app/styles`, где помимо стилей можно хранить еще ассеты для этих стилей

[`thoughts:`](https://conventionalcomments.org/) app, конечно, толком не регламентируется - но кажется это может упростить структуру

![image](https://user-images.githubusercontent.com/42924400/149591085-6d42be13-3032-4472-89ba-95caded340a5.png)

**Comment by azinit** [comment]
## 🟥 Теперь по критам

[`issue(public-api):`](https://conventionalcomments.org/) Прям ключевое что проявляется прям везде - это полное отсутствие декларации Public API и хотя бы минимально необходимых файлов реэкспортов

> Про линтер в конце еще скажу 😏 

![image](https://user-images.githubusercontent.com/42924400/149591483-379b7ffd-c98b-4e57-9c2d-8f8dcddf7f1b.png)

Хотя базовые и есть
![image](https://user-images.githubusercontent.com/42924400/149591553-efca9b1a-dd7c-4b3f-b615-9ce45480d57f.png)

[`nitpick(public-api):`](https://conventionalcomments.org/) Поэтому для большей устойчивости модулям и более удобного использования - можно объявить файлы реэкспорты в заявленных местах
(хотя бы на уровне слайсов и сегментов)

**Reply by azinit** [reply]
Вот тут еще примеры наглядные)

![image](https://user-images.githubusercontent.com/42924400/149591683-63045e83-2725-4f34-971b-e04fbe00a777.png)

![image](https://user-images.githubusercontent.com/42924400/149591813-ba240776-4923-4702-9e87-547c0dc4acd2.png)

![image](https://user-images.githubusercontent.com/42924400/149591819-f89cda95-0d74-440b-8277-ec8752aea72e.png)

![image](https://user-images.githubusercontent.com/42924400/149591828-7fffaaf0-6607-40a0-8ff9-75fe138fb25a.png)

**Reply by azinit** [reply]
А тут еще в целом непонятно - зачем рутовый компонент страницы выносить в canvas-page, если можно сделать общий `ui/index.tsx` 🤔 

![image](https://user-images.githubusercontent.com/42924400/149591935-4e120674-d827-47e2-9624-85b001ff5a77.png)

![image](https://user-images.githubusercontent.com/42924400/149591997-d2e83494-7824-4e24-9530-0b98e94db26d.png)

**Reply by azinit** [reply]
вот тут прям хорошо - при всех вопросах к неймингу внутри сегментов, как минимум есть файл реэкспорта, который помогает устойчивости и использованию
![image](https://user-images.githubusercontent.com/42924400/149594600-7652fa7a-f493-4cf9-82db-bb61a53e298a.png)

чего не скажешь про это

![image](https://user-images.githubusercontent.com/42924400/149594703-978cdb4e-7409-447c-b3e4-80eba435acea.png)

**Comment by azinit** [comment]
[`issue(layers):`](https://conventionalcomments.org/) Прям оч пухлая страница получается, даже несмотря на то, что это основная и единственная страница приложухи

![image](https://user-images.githubusercontent.com/42924400/149591880-d265760e-780b-4a96-9157-ffe61ba87db0.png)

**Reply by azinit** [reply]
[`suggestion(layers):`](https://conventionalcomments.org/) Как вариант - тут вот это прям просится, для выделения хотя бы в одну фичу
(если не группы фич)

> Не исключено даже, что тула это еще одна сущность, но мне кажется надо подождать развития сервиса - т.к. пока звучит чуть упорото 🤔 

![image](https://user-images.githubusercontent.com/42924400/149592051-c4d755d6-b138-41d6-8a68-252218423c5d.png)

**Reply by azinit** [reply]
[`suggestion(layers):`](https://conventionalcomments.org/) Еще как вариант - часть упомянутых компонентов в `pages/canvas/ui/**`  вынести в полноценные виджеты

Если не ради переиспользования, то хотя бы для упрощения логики страницы и поддерживания самих виджетов
(особенно если они слишком сильно подвязываются на страницу, можно плюс вместе с этим забрать часть логики модели/либ, если это будет в виджете отдельном)

**Comment by azinit** [comment]
[`issue(model, ui):`](https://conventionalcomments.org/) Прям типичная проблема - что очень сложно определить, где логика модели, а где ui, и что вообще такое lib
(иногда прям все в компоненте накидывают тупо, но у тебя есть хотя бы базовое разделение - уже хорошо)

Но тут мне кажется проблемы не совсем верного определения сегментов для такой логики - т.к. это явно логика "модели", т.е. бизнес логика и дата-флоу модуля

А в `lib` же мы обычно заносим чисто утилитарную логику, которая НЕ относится напрямую к UI-логике или дата-флоу
(т.е. туда можно складировать что-то общее, на что может подвязаться ui/model/api или другой модуль, но напрямую lib про них знать не должен, и тем более забирать у них их ключевую зону ответственности)

![image](https://user-images.githubusercontent.com/42924400/149592322-a36c8018-2bb0-4b4a-9233-78ef1c42b51d.png)

![image](https://user-images.githubusercontent.com/42924400/149592329-d12d99c5-bc83-419e-bd37-3b0d3a56a94b.png)

**Reply by azinit** [reply]
[`suggestion(model):`](https://conventionalcomments.org/) Вот тут кста прям наглядный пример-предложение как с таким можно поступать, без упарываний
(ты в целом в верном направлении мыслил)

https://github.com/feature-sliced/cra-template/pull/9#discussion_r780659063

**Reply by azinit** [reply]
[`thoughts:`](https://conventionalcomments.org/) Вот тут например можно прям сделать хук с `useState + handle` и вынести в модель, но некрит

![image](https://user-images.githubusercontent.com/42924400/149592830-9a14c333-4224-48e7-850f-baa3413f398e.png)

**Reply by azinit** [reply]
[`nitpick:`](https://conventionalcomments.org/) Еще пример 
С одной стороны - canvas действительно относится к UI, и это как раз хороший пример, когда у нас UI не только компонентами един

А с другой - тут прям видно, что происходит оч много манипуляций с данными - что явно можно снова вынести в хук и в модель

\> вся дата-логика будет в модели
\> а ui останется простым, к чему мы и стремимся
![image](https://user-images.githubusercontent.com/42924400/149594725-92a7d27d-64c3-46d9-a083-1a60188babb4.png)

**Comment by azinit** [comment]
[`issue(entities, features):`](https://conventionalcomments.org/) Тут вот верно с какой-то стороны, что сущностью определен `color`
(`color-picker` не совсем про это, но не суть)

Но фичи же - это как раз то ЧТО МЫ МОЖЕМ ДЕЛАТЬ С СУЩНОСТЬЮ

Т.е. в таком случае у нас color-picker больше походит на фичу, чем на часть сущности

![image](https://user-images.githubusercontent.com/42924400/149592913-327afd05-f574-4039-8e41-41cd5486bdab.png)

**Reply by azinit** [reply]
[`thoughts:`](https://conventionalcomments.org/) Т.е. если у нас еще будут еще юзкейсы, завязанные на цвет - то это + еще фичи, а entities будет унифицированно отдельно лежать

С другой стороны... Если щас color только в color-picker используется, то может не стоит торопиться так с выводами, и оставить `color-picker` как фичу, чтобы не было излишне раздроблено
(хотя тут скорее в целом нет до конца уверенности, что `color` это сущность точно должна быть)

**Comment by azinit** [comment]
[`issue(lib):`](https://conventionalcomments.org/) Любимая тема @sergeysova ))

TLDR: Либы (особенно если говорим про shared), должны проектироваться так, чтобы их можно было рассматривать как отдельные полноценные и самостоятельные решения, покрывающие определенный скоуп проблемы, и которые можно отдельно протестировать

> Более детально, можно прочитать тут:
> https://sova.dev/ru/why-utils-and-helpers-is-a-dump/

![image](https://user-images.githubusercontent.com/42924400/149593841-5e85d7d6-1ebb-4e5c-b23f-1e8aafade616.png)

**Reply by azinit** [reply]
[`suggestion(lib):`](https://conventionalcomments.org/) Конкретно тут - это то, что `generateSymbols` - это не либа)
`random.ts`, `string.ts`, `number.ts`  - еще да, но не первое

Может там пока и одна функция, но это оч редко когда повод точно так же называть либу, и уж тем более хранить по либе на каждую мини-функцию

![image](https://user-images.githubusercontent.com/42924400/149593204-4de246f0-706e-4507-a338-54095e4bfa5d.png)

**Reply by azinit** [reply]
[`suggestion(lib):`](https://conventionalcomments.org/) Тут же в цеолм сомнения, что это относится напрямую к useImageId а не условному `usePathname` (который в целом есть сомнения, что стоит выносить ради такого)

С другой стороны, если же хочется это унифицировать именно для imageId - то стоит такое хранить в либе рядом с местом использования
(чаще всего для этого всегда еще сущность отдельная находится - но кажется не совсем наш случай)

![image](https://user-images.githubusercontent.com/42924400/149593624-985c780e-4f44-4ef0-8fef-baeddae3fa0a.png)

**Reply by azinit** [reply]
[`suggestion(lib):`](https://conventionalcomments.org/) Тут прям хороший пример хука для либы, но опять же - нельзя выделить либу `hooks`

То что тут на скрине, это скорее либа `forms.ts`, или же даже положить рядом с условным инпутом в `shared/ui`

![image](https://user-images.githubusercontent.com/42924400/149593729-ad422018-8377-4d85-a2c6-30da1bd8f02b.png)

**Comment by azinit** [comment]
## 🟨 Теперь non-blocking (близки к концу уже :D)

[`nitpick(structure, kebab-case, non-blocking):`](https://conventionalcomments.org/) Очень большой спор, что лучше `PascalCase`, `kebab-case`, ~~полное отключение FSD в РФ~~

Но как уже говорил до этого в другом дискашне - кажется, что если выбираешь один способ описывания структуры, то лучше придерживаться его везде

**Чем проще - тем лучше**

Если же по тем или иным причинам начинаются встречаться в структуре в наименовании разные варианты - то это оч быстро приведет к хаосу в проекте в этом аспекте  (см. "теория разбитых окон")

Тут наглядные примеры - где желательно пофиксить
(много где kebab-case, но там где красное - прям выбивается из общей массы)

![image](https://user-images.githubusercontent.com/42924400/149594156-8e2ab690-311f-48c0-99f6-fd5a32936826.png)

![image](https://user-images.githubusercontent.com/42924400/149594166-6096bdbe-654b-4ecb-9c29-40d321811de2.png)

**Comment by azinit** [comment]
[`nitpick(non-blocking):`](https://conventionalcomments.org/) Тут прям сомнительно, что такие типы надо хранить на уровне app

Это ведь не только там используется, поэтому лучше хранить сразу где-то в `src/***.d.ts` файлах

![image](https://user-images.githubusercontent.com/42924400/149594373-6cf9f455-7cb1-4211-8bac-a1f0c256e32f.png)

**Comment by azinit** [comment]
[`nitpick(non-blocking):`](https://conventionalcomments.org/) Тут опять же сомнительно, что стоит роуты прям в shared/router хранить
~~(даже линтер ругается :D)~~, но прям назвать это критичным сложно

Лучше просто еще раз оставлю эту статью

https://feature-sliced.design/docs/concepts/issues/routes

![image](https://user-images.githubusercontent.com/42924400/149594430-fa4409f2-dced-42ad-a3cf-fd86482c34f0.png)

**Comment by azinit** [comment]
[`nitpick(non-blocking, shared-config):`](https://conventionalcomments.org/) тут не столько слайсы, сколько то, что в shared/config неплохо прорабатывать как раз не только "распаковку" значений из ENV, но и добавлять тут дефолтные значения и/или подстраховку на случай если переменная не найдена

Можно глянуть это как один из вариантов
https://github.com/ani-team/github-client/blob/workshop/feature-sliced-next/src/shared/config/index.ts
![image](https://user-images.githubusercontent.com/42924400/149594966-103b8d0a-258c-4e10-9ddb-accc3dc17d05.png)

**Comment by azinit** [comment]
[`nitpick(non-blocking):`](https://conventionalcomments.org/) Первое - про конвенциональность нейминга

Если хотим придерживаться какого-то +- стандарта, то лучше назвать директорию такую `shared/assets` (как ты уже делал), либо же сразу положить это в `shared/ui`


![image](https://user-images.githubusercontent.com/42924400/149595115-34ec66dd-4880-45da-a450-70d572aed757.png)

**Reply by azinit** [reply]
Другой момент - это насколько нужно хранить иконки унифицированно в одном shared

Оч часто хватает того, чтобы иконки класть с местом использования, но чтобы не размусоливать - лучше оставлю тут тред

https://t.me/feature_sliced/1742

**Comment by azinit** [comment]
[`nitpick(non-blocking, model):`](https://conventionalcomments.org/)  Лучше логику связанную с моделью - хранить прям в модели
Когда что-то описано в компоненте страницы, что-то в другом компоненте - прям проблемно становится это изучать и находить

Т.е. лучше эти вотчи опложить ближе к модели страницы

> Ну и в контексте эффектора "вотчить" не совсем бест практис

![image](https://user-images.githubusercontent.com/42924400/149595524-119a137e-be9b-4c9f-8cf2-7467b6601652.png)

**Reply by azinit** [reply]
Тут же прям модель похожа на ту, что описывает методология, да)
(только ругается на либу из-за отсутствия public-api)

![image](https://user-images.githubusercontent.com/42924400/149595616-92e5681d-bf6a-4274-8c3f-3312a2006167.png)

**Reply by tednaaa** [reply]
Я просто думал что если мы юзаем другой сегмент из того же слайса, то не нужно публик апи =)

**Reply by azinit** [reply]
Ну если выделять по критичности,то скорее так:
(от наиболее важного, к наименее)

Public API: 
=> component/store reexport (general reexport)
=> slice reexport
=> segment reexport
=> inner impl module reexport

**Comment by azinit** [comment]
## 🔳 Теперь чисто мысли

[`thoughts:`](https://conventionalcomments.org/) Тут надеюсь не хочется создать "гиперконтейнер, который возьмет на себя всю ответственность по работе с данными что только можно", а чисто компонент для отображения))

![image](https://user-images.githubusercontent.com/42924400/149595707-1e549ba0-8682-4984-9960-f1ee01b35d74.png)

**Comment by azinit** [comment]
[`thoughts(canvas, ui):`](https://conventionalcomments.org/) Тут как уже говорил - canvas вроде как история чисто про UI

Однако тут видно, что есть обработчики и базовая логика переключения - что действительно лучше в модели смотрится

выглядит в целом норм, просто немного флешбечит, что сильно много специфики UI в модели

![image](https://user-images.githubusercontent.com/42924400/149595770-4eb4675e-9aa2-40c9-b636-9a5a462e9428.png)

**Reply by azinit** [reply]
Хотя мб @GhostMayor @Postamentovich будет что тут сказать по этому поводу 🤔

**Comment by azinit** [comment]
## FINAL

**ФУХ! Боюсь смотреть сколько получилось сообщух, но постарался все основное расписать)**

Спасибо за проект для ревью, постарался где мог сгруппировать схожие проблемы -надеюсь облегчит восприятие

**Reply by azinit** [reply]
@tednaaa вот тут так же предлагаю принять участие в "закрытом" бета-тесте нашего еслинт конфига

Многие варнинги со скринов - как раз на основании его выводились)
> Что странно, прям не было никаких проблем в плане нарушения [правил импортов между слоями/слайсами](https://github.com/feature-sliced/eslint-config/tree/master/rules/layers-slices), но возможно потому что их немного

https://www.npmjs.com/package/@feature-sliced/eslint-config
https://www.npmjs.com/package/@feature-sliced/eslint-plugin-messages

**Reply by azinit** [reply]
@tednaaa Было бы круто довести это все дело до полноценного примера FSD, дай знать если захочется)
